src/pybind/mgr/cephadm/services/ingress.py

```diff
@@ -89,7 +89,23 @@ def haproxy_generate_config(
 
         if backend_spec.service_type == 'nfs':
             mode = 'tcp'
-            by_rank = {d.rank: d for d in daemons if d.rank is not None}
+            ranked_daemons = [d for d in daemons if (d.rank is not None and d.rank_generation is not None)]
+            by_rank: Dict[int, DaemonDescription] = {}
+            for d in ranked_daemons:
+                # It doesn't seem like mypy can figure out that rank
+                # and rank_generation for both the daemon we're looping on
+                # and all those in by_rank cannot be None due to the filtering
+                # when creating the ranked_daemons list, which is why these
+                # seemingly unnecessary assertions are here.
```

Comment:
did you try to put an explicit type? something like `List[Daemon]` .. maybe this would force mypy to understand that antries count no be None!

Reply:
I think my comment might not have been totally clear. It's not the daemon itself that it thinks could be `None`, it's specifically the `rank` and `rank_generation` attributes of the daemons it thinks could be `None`. 

Afaik there's no way to tell mypy that all objects in this list will not have a certain value for one of their attributes (or even just whether they can be `None` or not). Except maybe making a new type that is a `DaemonDescription` but with those attributes not being optional and converting the `DaemonDescription` objects to that type as the list is constructed. But that sounds more painful than just doing the assertions.

Reply:
Agreed, the earlier filter is a filter by value and not a filter by type, and so since the type still allows Nones in these places mypy will do as it does. I'm OK with the assertions (as the mypy docs note this is a valid way to do [inline narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html#type-narrowing-expressions).  If it got really annoying & tedious it might be helpful to create a local class (or something like a namedtuple) - but I don't think it is strictly necessary in this case.

Reply:
@adk3798 you are right and your comment in the code is clear (now that I re-read it for the second time).